units/cache_historical_test_data.go

```diff
@@ -0,0 +1,289 @@
+		j.AddError(errors.Wrap(err, "error finding tasks to update"))
+		return
+	}
+
+	generateMap := generateFunctions{
+		HourlyFns: map[string]generateStatsFn{
+			"test": stats.GenerateHourlyTestStats,
+		},
+		DailyFns: map[string]generateStatsFn{
+			"test": stats.GenerateDailyTestStatsFromHourly,
+			"task": stats.GenerateDailyTaskStats,
+		},
+	}
+
+	err = updateHourlyAndDailyStats(projectId, statsToUpdate, jobTime, generateMap, tasksToIgnore)
+	if err != nil {
+		j.AddError(errors.Wrap(err, "error generating hourly test stats"))
+		return
+	}
+
+	// update last sync
+	err = stats.UpdateStatsStatus(projectId, jobTime, syncToTime)
+	if err != nil {
+		j.AddError(errors.Wrap(err, "error updating last synced date"))
+		return
+	}
+
+	grip.Info(message.Fields{
+		"job_id":     j.ID(),
+		"project_id": projectId,
+		"message":    "job completed",
+	})
+}
+
+func getTasksToIgnore(projectId string) ([]*regexp.Regexp, error) {
+	ref, err := model.FindOneProjectRef(projectId)
+	if err != nil {
+		return nil, err
+	}
+
+	filePatternsStr := ref.FilesIgnoredFromCache.FilePatterns
+
+	return splitPatternStringToRegexList(filePatternsStr)
+}
+
+func splitPatternStringToRegexList(filePatternsStr string) ([]*regexp.Regexp, error) {
+	filePatterns := strings.Split(filePatternsStr, ",")
+
+	var tasksToIgnore []*regexp.Regexp
+	for _, patternStr := range filePatterns {
+		pattern := strings.Trim(patternStr, " ")
+		if pattern != "" {
+			regexp, err := regexp.Compile(pattern)
+			if err != nil {
+				grip.Warning(message.Fields{
+					"message": fmt.Sprintf("Could not compile regexp from '%s'", pattern),
+				})
+				return nil, err
+			}
+			tasksToIgnore = append(tasksToIgnore, regexp)
+		}
+	}
+
+	return tasksToIgnore, nil
+}
+
+func updateHourlyAndDailyStats(projectId string, statsToUpdate []stats.StatsToUpdate,
+	jobTime time.Time, generateFns generateFunctions, tasksToIgnore []*regexp.Regexp) error {
+	for name, genFn := range generateFns.HourlyFns {
+		err := iteratorOverHourlyStats(statsToUpdate, jobTime, genFn, name, tasksToIgnore)
+		if err != nil {
+			return err
+		}
+	}
+
+	dailyStats := buildDailyStatsRollup(statsToUpdate)
+
+	for name, genFn := range generateFns.DailyFns {
+		err := iteratorOverDailyStats(projectId, dailyStats, jobTime, genFn, name, tasksToIgnore)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func iteratorOverDailyStats(projectId string, dailyStats dailyStatsRollup, jobTime time.Time,
+	fn generateStatsFn, displayName string, tasksToIgnore []*regexp.Regexp) error {
+	for day, stats := range dailyStats {
+		for requester, tasks := range stats {
+			taskList := filterIgnoredTasks(tasks, tasksToIgnore)
+			if len(taskList) > 0 {
+				err := fn(projectId, requester, day, taskList, jobTime)
+				if err != nil {
+					grip.Info(message.Fields{
+						"project_id": projectId,
+						"sync_date":  day,
+						"job_time":   jobTime,
+						"message":    fmt.Sprintf("error syncing daily %v stats", displayName),
```

Comment:
I'd avoid sprintf'ing here unless you really need to. I'd just add display name to the fields, and have message me a string.

Reply:
updated.